docs: copy best practices from Juju docs

tonyandrewmeyer opened the pull request:
This PR migrates best practices from [the Juju docs](https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/charm/charm-development-best-practices/), where the best practice applies to Charmcraft. It is paired with a [similar PR for ops](https://github.com/canonical/operator/pull/1615).

The intention is to standardise best practices so that the docs can be scraped to collect best practices across Ops, Charmcraft, juju, and elsewhere in the charming ecosystem, to provide a comprehensive checklist for charmers.

My suggestion is for ReST we use:

```
.. admonition:: Best practice
    :class: hint

    Content goes here
```

This provides an admonition box that has the title "Best Practice", but is otherwise styled as a hint (David Wilding is working on putting a custom icon into the starter pack, to replace the ? icon that Furo defaults to).

The best practices here were selected in a review by @tmihoc and @dwilding and myself, although all problems with the suggested wording are my fault alone.

erinecon committed:
docs: maintain Django tutorial (#2196)

Some specific fixes to the Django tutorial:
* Don't need to explicitly have `python3 manage.py`. Django developers
are more used to `./manage.py` instead. I have updated the tutorial and
`task.yaml` file.
* Any reference to `/django-hello-world` in text has been updated to
`~/django-hello-world`.
* Updated indentation in a code snippet that was rendering incorrectly.

Additional reviewers: @jdkandersson

alithethird committed:
feat(smtp): Add SMTP integration to 12 Factor (#2199)

This PR is related to 12 Factor [SMTP
PR](https://github.com/canonical/paas-charm/pull/32).

## Rationale

Some applications require an SMTP integration to send emails to their
clients. This PR is a follow-up to the latest [12 Factor
PR](https://github.com/canonical/paas-charm/pull/32) for adding SMTP
integration.

## Changes

- Add commented-out SMTP relation required objects to extension
charmcraft.yaml.j2 files.
- Add references about SMTP into extension reference documents.
- Add documentation about SMTP integration and generated environment
variables
- Add the smtp charm lib to the initializer.

Reviewers:
@jdkandersson
@javierdelapuente
@erinecon

---------

Co-authored-by: Erin Conley <erin.conley@canonical.com>

tonyandrewmeyer committed:
docs: copy best practices from Juju docs

tonyandrewmeyer commented:
(Still in draft - the intention is to move both PRs to review at the same time, hopefully later this week. Review before then is ok, but is probably best left until out of draft).

tmihoc commented on docs/howto/manage-charms.rst:
```diff
@@ -65,6 +66,34 @@ in, a charm name different from the name of the root directory, etc.
+    The `charming-actions <https://github.com/canonical/charming-actions>`_
+    repository includes actions to ensure that libraries are up-to-date, publish
+    charms and libraries, and more.
+
```
@lengau Would you say these two things we have here as tips are in fact best practices too?

tmihoc commented on docs/howto/manage-charms.rst:
```diff
@@ -65,6 +66,34 @@ in, a charm name different from the name of the root directory, etc.
+
+    The quality assurance pipeline of a charm should be automated using a
+    continuous integration (CI) system.
+
```
FWIW I do not think this is redundant -- I can definitely see it being useful to some of our intended target audience.

tmihoc commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -424,6 +432,46 @@ secret URI.
+.. admonition:: Best Practice
+    :class: hint
+
+    Charms should keep the user experience of the operator as simple and obvious
```
```suggestion
    Charms should keep the user experience of the operator as smooth
```

tmihoc commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -424,6 +432,46 @@ secret URI.
+    as possible. Aim to make the application deployable without providing any
+    further configuration options: ``juju deploy foo`` is preferable to
+    ``juju deploy foo --config required-option=value``.
+
```
You might streamline this as: To make the user's experience as smooth as possible, aim to...

tmihoc commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -424,6 +432,46 @@ secret URI.
+    Consider which of your application's configuration options you should
+    initially expose. If your application has many config options, it may be
+    prudent to provide access to a select few, and add support for more as the
+    need arises.
```
It feels like the intent here is to say: Be thrifty with how many configs = knobs you expose. Just like Juju, a charm is supposed to be an opinionated tool. I'd rephrase something like that. (Because "initially expose" makes me wonder: Is it okay to end up a gazillion configs later on?)

tmihoc commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -424,6 +432,46 @@ secret URI.
+.. admonition:: Best Practice
+    :class: hint
+
+    The description a string type. YAML supports two types of formats for that:
```
```suggestion
    The description is a string type. YAML supports two types of formats for that:
```

tmihoc replied:
I understand that this in the original Best practice doc this was about the config key. However, I feel like this is a comment on the description keys in charmcraft.yaml more generally, so many it's really a best practice at the level of the entire doc. Speaking of which, one more candidate for that top-level note might be https://discourse.charmhub.io/t/yaml-anchors-and-aliases/5683 , which we used to have in the charm SDK docs and I never ported into Charmcraft docs because I wasn't sure about the relevance. @lengau Can you please advise?

tmihoc commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -424,6 +432,46 @@ secret URI.
+  larger deployments, or "profile: ci" for limited resource usage during
+  testing.
+
 
```
@lengau Do we really want this?

lengau replied:
I think this is fine to keep, though it's probably a best practice as well.

tmihoc commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -567,6 +615,12 @@ for users) and the default is 0 (root).
+
+    Documentation links should apply to the charm, and not to the application that
+    is being charmed. Assume that the user already has basic competency in the use
+    of the application.
```
Note for the future: We'll want to add one more thing to this note, that is, links to examples of nicely documented charms.

tmihoc commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -592,6 +646,18 @@ determines the name administrators will ultimately use to deploy the charm. E.g.
+
+    Don't add an ``operator`` or ``charm`` prefix or suffix. For naming a
+    repository, see :ref:`initialise-a-charm`.
+
```
Thank you for including the negatives (which I forgot to include when I added this text originally).

tmihoc reviewed (approved):
Thank you so much for this work: It's very important work and I appreciate your speed and thoroughness. I've left some comments. None blockers -- just things we should try to sort out potentially.

tonyandrewmeyer commented on docs/howto/manage-charms.rst:
```diff
@@ -11,10 +11,11 @@ Manage charms
-     ``<charm name>-operator``. For the charm name, see :ref:`specify-a-name`.
+    If you're setting up a git repository: name it using the pattern
+    ``<charm name>-operator``. For the charm name, see :ref:`specify-a-name`.
 
```
I found the source - it's coming from [Furo](https://pradyunsg.me/furo/reference/admonitions/). Maybe "important" would be the best choice (avoiding using a custom style)? It's more of a blue-green and the icon is I think a flame.

tonyandrewmeyer replied:
The [ops PR](https://ops--1615.org.readthedocs.build/en/1615/howto/manage-stored-state.html) has an example of the style we're going for.

tonyandrewmeyer committed:
docs: add `charm-user` in the charmcraft.yaml file reference (#2195)

The `charm-user` field is documented in the metadata.yaml file
reference, but is missing from the charmcraft.yaml file reference, even
though it does work there. This PR adds it to that reference.

As the docs currently are, it gives the incorrect impression that you
cannot use the combined charmcraft.yaml file to specify `charm-user`,
and must have separate metadata, actions, and config files.

---------

Co-authored-by: Erin Conley <erin.conley@canonical.com>
Co-authored-by: Ali UĞUR <39213991+alithethird@users.noreply.github.com>

lengau committed:
feat(templates): change the templates to use platforms notation (#2200)

This updates the templates to use the `platforms` notation rather than
`bases`. It keeps them building on Jammy for the timebeing though.

CRAFT-4150

Fixes #2167

erinecon committed:
docs: add structure to 12-factor how-to guides (#2197)

Currently the 12-factor how-to guides were split into two documents: one
for the async configuration, and another containing all other how-to
sections. This pull request consolidates the two documents into a
directory, `manage-12-factor-charms`, with added structure by splitting
up the sections into separate pages.

### Summary of changes
* Created `manage-12-factor-charms` directory with an index page and
four different categories: Initialization, Configuration, Integration,
and Usage
* Consolidated the flask-async document into the Configuration category 
* Added redirects

Additional reviewers: @jdkandersson

---------

Co-authored-by: Michael DuBelko <michael.dubelko@gmail.com>
Co-authored-by: Alex Lowe <alex.lowe@canonical.com>
Co-authored-by: Ali UĞUR <39213991+alithethird@users.noreply.github.com>
Co-authored-by: Tony Meyer <tony.meyer@canonical.com>

tonyandrewmeyer committed:
docs: be more specific about the -k8s suffix, per the discussion in Discourse

tonyandrewmeyer committed:
docs: add a best practice note for the optional field as per Discourse

tonyandrewmeyer committed:
Merge branch 'main' into add-best-practice-notes

tmihoc commented on docs/howto/manage-charms.rst:
```diff
@@ -11,10 +11,11 @@ Manage charms
 
-    **Best practice:** If you're setting up a git repository: Name it on the pattern
-     ``<charm name>-operator``. For the charm name, see :ref:`specify-a-name`.
+    If you're setting up a git repository: name it using the pattern
```
```suggestion
    If you're setting up a `git` repository: name it using the pattern
```

tonyandrewmeyer committed:
docs: put backticks around 'git'

Co-authored-by: Teodora Mihoc <teodora.mihoc@canonical.com>

sergiusens reviewed (dismissed):
I am going to discuss this with @evildmp next week, please be patient.

tonyandrewmeyer committed:
docs: expand the information about the provided tooling, and add two anchors that ops needs.

tonyandrewmeyer commented:
> I am going to discuss this with @evildmp next week, please be patient.

Hi @sergiusens - any update on this?

sergiusens reviewed (commented):
I will abstain my review. These docs however seem to belong more on the Charmhub side than Charmcraft.

evildmp commented:
There seems to be real doubt about this. If the engineering manager isn’t happy about a PR then I think it’s right that we make another effort to understand the situation all together and agree on some way forward. Sorry to hold up the work, but I think we need to do that first.

lengau commented on docs/howto/manage-resources.rst:
```diff
@@ -79,6 +79,11 @@ will be a new revision.
 
     See more: :ref:`ref_commands_upload-resource`
 
+.. admonition:: Best practice
```
Per today's meeting: I think this is worth keeping in charmcraft *and* adding to ops :-)

tonyandrewmeyer replied:
As per the above, I've left this in, and also added it in the PR for ops that include things cur from here.

medubelko commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -63,6 +63,14 @@ The value of this key is the contents of :ref:`actions-yaml-file`.
         :start-at: actions:
         :end-before: analysis:
 
+.. admonition:: Best practice
+    :class: hint
```
I don't think we need the admonitions. They feel like visual noise to my eyes.

tonyandrewmeyer replied:
There are 3 reasons for these:
* We want to really call attention to these, because they are the markers of what we consider mature/quality charms.
* We want to have a common look for best practice across the charming ecosystem, so that even though the docs are decentralised they feel like parts of a common platform.
* We want to have the best practices in the places that they'll most likely be found by charmers, rather than in one big list like Juju used to have, but we also need to be able to automatically collect them all (for example, for checking in a quality review process). For this, it's not important what they look like, but collecting them across all the tools is much simpler if there's a standard (or rather two: one for Markdown and one for ReST) rather than having to implement collectors for each tool.

medubelko commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -63,6 +63,14 @@ The value of this key is the contents of :ref:`actions-yaml-file`.
+.. admonition:: Best practice
+    :class: hint
+
+    Prefer lowercase alphanumeric names, separated with dashes if required. For
```
```suggestion
    Prefer lowercase alphanumeric names, and use hyphens (-) to separate words. For
```

medubelko commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -460,6 +468,45 @@ secret URI.
+.. admonition:: Best practice
+    :class: hint
+
+    Make the user experience of the operator as smooth as possible. The
+    application should be deployable without providing any further configuration
+    options: ``juju deploy foo`` is preferable to
+    ``juju deploy foo --config required-option=value``.
+
+.. admonition:: Best practice
+    :class: hint
+
+    Just like Juju, a charm is an opinionated tool. Configure the application
+    with the best defaults, and only expose application configuration options
+    when necessary.
```
Do you think we could combine these points? They feel almost the same to me. I prefer the copy in the second paragraph, at any rate.

medubelko commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -460,6 +468,45 @@ secret URI.
+.. admonition:: Best practice
+    :class: hint
+
+    The description is a string type. YAML supports two types of formats for
+    that: ``block scalar`` and ``flow scalar``. Prefer the block style
+    (using ``|``) to keep newlines. Using ``>`` will replace newlines with
+    spaces and make the result harder to read on Charmhub.
```
I would cut this. It's more like a nice-to-know for a how-to guide than essential text in a reference, in my opinion.

medubelko commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -460,6 +468,45 @@ secret URI.
+
+.. tip::
+
+  For very complex applications, consider providing "configuration profiles",
+  which can group values for large configs together. For example,
+  "profile: large" that tweaks multiple options under the hood to optimise for
+  larger deployments, or "profile: ci" for limited resource usage during
+  testing.
```
```suggestion
  For very complex applications, consider providing configuration profiles,
  which can group values for large configs together. For example,
  a ``profile: large`` that tweaks multiple options under the hood to optimize for
  larger deployments, or a ``profile: ci`` for limited resource usage during
  testing.
```

It's been slow going, but we have been creeping toward US English.

medubelko commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -460,6 +468,45 @@ secret URI.
+    (using ``|``) to keep newlines. Using ``>`` will replace newlines with
+    spaces and make the result harder to read on Charmhub.
+
+.. tip::
```
Can we make this a regular paragraph? :eyes:

medubelko commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -628,6 +681,22 @@ determines the name administrators will ultimately use to deploy the charm. E.g.
+    hyphens) and follow the pattern ``<workload name in full>[<function>][-k8s]``.
+    For example, ``argo-server-k8s``.
+
+    Include the '-k8s' suffix on all charms that run on a Kubernetes cloud,
```
```suggestion
    Include the ``-k8s`` suffix on all charms that run on a Kubernetes cloud,
```

medubelko commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -628,6 +681,22 @@ determines the name administrators will ultimately use to deploy the charm. E.g.
+    unless the charm has no workload or you know that there will never be
+    a machine version of the charm.
+
+    Don't include an organisation or publisher in the name.
```
```suggestion
    Don't include an organization or publisher in the name.
```

medubelko reviewed (commented):
Following up on my commitment to review the `charmcraft.yaml` reference.

CC @tonyandrewmeyer

tonyandrewmeyer committed:
docs: remove some content that can be added to ops instead.

tonyandrewmeyer committed:
docs: apply suggestions from code review

Co-authored-by: Michael DuBelko <michael.dubelko@gmail.com>

tonyandrewmeyer committed:
docs: changes per review comments.

tonyandrewmeyer commented on docs/howto/manage-charms.rst:
```diff
@@ -11,10 +11,11 @@ Manage charms
 ------------------
 
-.. note::
+.. admonition:: Best practice
```
This is in the how-to but it doesn't change the content, it just adjusts it to be a best practice rather than only a note.

tonyandrewmeyer commented on docs/howto/manage-charms.rst:
```diff
@@ -89,14 +91,6 @@ to ``charm`` in your project file:
 Specify a name
 ~~~~~~~~~~~~~~
 
-.. note::
```
Note that this isn't removed, just moved, as the expectation is that charmers are most likely to notice it when using the reference rather than the how-to.

tonyandrewmeyer committed:
Merge branch 'main' into add-best-practice-notes

tonyandrewmeyer commented:
Per the discussion last week, I've removed most of the how-to content leaving the changes mostly confined to the `charmcraft.yaml` reference. I've also made the adjustments suggested by @medubelko.

medubelko reviewed (approved):
Looks great, thanks @tonyandrewmeyer!

lengau commented on docs/howto/manage-charms.rst:
```diff
@@ -11,10 +11,11 @@ Manage charms
 
-    **Best practice:** If you're setting up a git repository: Name it on the pattern
-     ``<charm name>-operator``. For the charm name, see :ref:`specify-a-name`.
+    If you're setting up a `git` repository: name it using the pattern
```
```suggestion
    If you're setting up a ``git`` repository: name it using the pattern
```

lengau committed:
fix: minor formatting

lengau committed:
Merge branch 'main' into add-best-practice-notes

lengau commented on docs/howto/manage-resources.rst:
```diff
@@ -79,6 +79,12 @@ will be a new revision.
+.. admonition:: Best practice
+    :class: hint
+
+    For resources that are binary files, provide binaries for all the CPU 
```
```suggestion
    For resources that are binary files, provide binaries for all the CPU
```

lengau commented on docs/reference/files/charmcraft-yaml-file.rst:
```diff
@@ -460,6 +468,31 @@ secret URI.
+
+.. admonition:: Best practice
+    :class: hint
+    
```
```suggestion

```

lengau committed:
fix: lint
